### ISSUE TITLE
[question] test_requires conflic with requires - expected behavior?

### ISSUE THREAD

**Issue by actual-daniel** [conan-io/conan#13286]
### What is your question?

I tried to find a answer to this in the documentation and in other issues but was not able to - most issues are based on misuse of the `build_requires`/`test_requires` feature.

Lets say we have the following graph: 
![](https://raw.githubusercontent.com/actual-daniel/conan-test-requires/main/graph.drawio.svg)

* `A`: Some common library
* `B`: A library which uses `A` as normal requirement i.e. it is part of the package content
* `C`: Uses `B` as normal requirement and uses `A` as test requirement because it is only used for some tests which are not part of the package content
* `X`: Consumer of `C` which overrides the dependencies onto A

This works as long a binary package of `C` is already provided. But if `C` has to be built (e.g. because different profiles are used) things fall apart:
```
ERROR: Conflict in PkgC/1.0.0:
    'PkgC/1.0.0' requires 'PkgA/1.0.0' while 'PkgB/1.0.0' requires 'PkgA/2.0.0'.
    To fix this conflict you need to override the package 'PkgA' in your root package.
```

I created a small test case which shows the situation: https://github.com/actual-daniel/conan-test-requires

Is this expected? I am currently not sure how to solve this. One way would be to just use a normal `requires` from `C` onto `A` even though I feel like this is semantically not correct. 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @actual-daniel 

Thanks for your question.
Yes, that is definitely a conflict, let me illustrate with some realistic example:

- We have ``gtest/1.0`` test framework library that depends (normal ``require``) on ``zlib/1.2.11``
- We have ``pkg/1.0``, a normal library, that ``test_requires("gtest/1.0")``, and it also depends, normal ``require`` ``zlib/1.2.13``

When we are building a final ``test1.exe`` that test some functionality of ``pkg``, it will need to transitively link with ``zlib/1.2.13``, because it is used by ``pkg``. But the ``test1.exe`` also depends on ``gtest/1.0`` that in turns depends on ``zlib/1.2.11``.
All of these are in the "host" context of "pkg" and "test1". So both zlib different versions will definitely conflict, as we cannot link an executable that depends on 2 different versions of the same library. At the end of the day, the test framework is no different than a regular library, the only difference is that its usage is not propagated downstream to consumers of ``pkg``, but the conflict in package already happens.

To avoid it, you need to converge the versions, and naturally the one to be respected is the one that will be used by production code, so the ``test_requires`` should adapt to that.

**Comment by actual-daniel** [comment]
@memsharded - Thanks for your answer!

> To avoid it, you need to converge the versions, and naturally the one to be respected is the one that will be used by production code, so the test_requires should adapt to that

But this means we have to define a common version for the whole dependency tree right? For my use case this is not really possible because the package is used by different teams working  and updating their dependencies independently. 

So I guess the easies fix would be to change the `test_requires` to a normal `requires`?

**Comment by memsharded** [comment]
> But this means we have to define a common version for the whole dependency tree right? 

The common version for the whole dependency graph still needs to happen. All the regular linked libraries cannot have different versions. So it is basically matching the ``test_requires`` to the general graph requires. Maybe using the ``test_requires`` with a version-range helps if the problem is that the dependency version keeps evolving.


Also, please take into account that in 2.0, override for conflict resolution should be explicit, and in general it is better to not rely on overriding, but on version ranges or on synchronized/converging exact versions.

**Comment by actual-daniel** [comment]
I was not able to resolve this using version ranges in conan 1.59. It gets resolved correctly in conan 2.0 though. 

Guess we workaround the issue until we migrate to conan 2. Thanks for your help!

**Comment by memsharded** [comment]
Ok, thanks for the feedback.
So if there is some workaround like converting to ``requires`` and a path forward for improvement, working in 2.0, I think we can close this ticket. Thanks!
